### DISCUSSION TITLE
typing for httpx.ASGITransport does not allow starlette applications

### DISCUSSION THREAD

**Discussion by RommelLayco** [encode/httpx#2832] [Potential Issue]
The typing for the app field in `httpx.ASGITransport`  was changed https://github.com/encode/httpx/pull/2468.

Previously the type of the app field was a `Callable` now it is

https://github.com/encode/httpx/blob/053bc57c3799801ff11273dd393cb0715e63ecf9/httpx/_transports/asgi.py#L22-L24

This causes apps that uses starlette to fail the type check when using mypy.

# Steps to reproduce

1. Create this script
```python
from starlette.applications import Starlette
from starlette.responses import HTMLResponse
from starlette.routing import Route

import httpx


async def hello(request):
    return HTMLResponse("Hello World!")


app = Starlette(routes=[Route("/", hello)])

transport = httpx.ASGITransport(app=app, client=("1.2.3.4", 123))
```
*NB* taken from https://www.python-httpx.org/async/#calling-into-python-web-apps

2. Run mypy against the above script

`mypy example.py`

## Expected

mypy to pass no errors

## Actual

```
example.py:14: error: Argument "app" to "ASGITransport" has incompatible type "Starlette"; expected "Callable[[dict[str, Any], Callable[[], Awaitable[dict[str, Any]]], Callable[[dict[str, Any]], Coroutine[None, None, None]]], Coroutine[None, None, None]]"  [arg-type]
example.py:14: note: "Starlette.__call__" has type "Callable[[Arg(MutableMapping[str, Any], 'scope'), Arg(Callable[[], Awaitable[MutableMapping[str, Any]]], 'receive'), Arg(Callable[[MutableMapping[str, Any]], Awaitable[None]], 'send')], Coroutine[Any, Any, None]]"
```

**Comment by YDX-2147483647** [comment]
I can reproduce this with starlette 0.31.1 and httpx 0.25.0.

The key is that [`httpx._transports.asgi._Send` needs `Dict[str, Any]`](https://github.com/encode/httpx/blob/053bc57c3799801ff11273dd393cb0715e63ecf9/httpx/_transports/asgi.py#L17-L21C1), but [`starlette.types.Message` gives `MutableMapping[str, Any]`](https://github.com/encode/starlette/blob/93074e82ab15e860456d4a7074c0494eed78ea0c/starlette/types.py#L11-L14).
`MutableMapping` also includes [collections.defaultdict](https://docs.python.org/3.11/library/collections.html#collections.defaultdict), [collections.OrderedDict](https://docs.python.org/3.11/library/collections.html#collections.OrderedDict) and [collections.Counter](https://docs.python.org/3.11/library/collections.html#collections.Counter), etc. according to [Python Doc](https://docs.python.org/3.11/glossary.html#term-mapping).

## Possible Fix

Loose `_Message` from `Dict` to `MutableMapping`, and annotate relevant interfaces as `_Message` (as [starlette does](https://github.com/encode/starlette/blob/93074e82ab15e860456d4a7074c0494eed78ea0c/starlette/types.py#L10-L16C28)).

https://github.com/encode/httpx/compare/master...YDX-2147483647:httpx:message-typing or https://github.com/YDX-2147483647/httpx/commit/be45c1376079458719ea93a6a371c2eec15e3469

> [!NOTE]
>
> Mypy is happy about that, but I have NOT checked other things. Therefore no PR for now.

**Comment by antonwnk** [comment]
looks like it's solved now

https://github.com/encode/httpx/issues/3111
